oauth_provider/oauth2/validator.py

```diff
@@ -0,0 +1,253 @@
+            ('code', '=', code),
+        ])
+        return redirect_uri == code.redirect_uri_id.name
+
+    def get_default_redirect_uri(self, client_id, request, *args, **kwargs):
+        """
+        Returns the default redirect URI for the client
+        """
+        client = http.request.env['oauth.provider.client'].search([
+            ('identifier', '=', client_id),
+        ])
+        return client.redirect_uri_ids and client.redirect_uri_ids[0].name \
+            or ''
+
+    def get_default_scopes(self, client_id, request, *args, **kwargs):
+        """
+        Returns a list of default scoprs for the client
+        """
+        client = http.request.env['oauth.provider.client'].search([
+            ('identifier', '=', client_id),
+        ])
+        return ' '.join(client.scope_ids.mapped('code'))
+
+    def get_original_scopes(self, refresh_token, request, *args, **kwargs):
+        """
+        Returns the list of scopes associated to the refresh token
+        """
+        token = http.request.env['oauth.provider.token'].search([
+            ('client_id', '=', request.client.id),
+            ('refresh_token', '=', refresh_token),
+        ])
+        return token.scope_ids.mapped('code')
+
+    def invalidate_authorization_code(
+            self, client_id, code, request, *args, **kwargs):
+        """
+        Invalidates an authorization code
+        """
+        code = http.request.env['oauth.provider.authorization.code'].search([
+            ('client_id.identifier', '=', client_id),
+            ('code', '=', code),
+        ])
+        code.write({'active': False})
+
+    def is_within_original_scope(
+            self, request_scopes, refresh_token, request, *args, **kwargs):
+        """
+        Check if the requested scopes are within a scope of the refresh token
+        """
+        token = http.request.env['oauth.provider.token'].search([
+            ('client_id', '=', request.client.id),
+            ('refresh_token', '=', refresh_token),
+        ])
+        return set(request_scopes).issubset(
+            set(token.scope_ids.mapped('code')))
+
+    def revoke_token(self, token, token_type_hint, request, *args, **kwargs):
+        """
+        Revoke an access of refresh token
+        """
+        db_token = http.request.env['oauth.provider.token'].search([
+            ('token', '=', token),
+        ])
+        # If we revoke a full token, simply unlink it
+        if db_token:
+            db_token.unlink()
+        # If we revoke a refresh token, empty it in the corresponding token
+        else:
+            db_token = http.request.env['oauth.provider.token'].search([
+                ('refresh_token', '=', token),
+            ])
+            db_token.refresh_token = False
+
+    def rotate_refresh_token(self, request):
+        """
+        Determine if the refresh token has to be renewed
+        Called after refreshing an access token
+        """
+        return True
```

Comment:
This is the implementation of an interface.
The default value in the interface is True, si this is not needed, but I preferred to implement all methods. We can remove this without any issue, if it is better.